frontend/src/components/datasources/datasources.tsx

```diff
@@ -0,0 +1,833 @@
+        }
+        icon={<DatabaseIcon />}
+      />
+    );
+  }
+
+  const hasSearch = !!searchValue.trim();
+
+  return (
+    <Command
+      className="border-b bg-background rounded-none h-full pb-10 overflow-auto"
+      shouldFilter={false}
+    >
+      <div className="flex items-center w-full">
+        <CommandInput
+          placeholder="Search tables..."
+          className="h-6 m-1"
+          value={searchValue}
+          onValueChange={(value) => {
+            // If searching, remove open previews
+            closeAllColumns(value.length > 0);
+            setSearchValue(value);
+          }}
+          rootClassName="flex-1 border-r"
+        />
+        {hasSearch && (
+          <button
+            type="button"
+            className="float-right border-b px-2 m-0 h-full hover:bg-accent hover:text-accent-foreground"
+            onClick={() => setSearchValue("")}
+          >
+            <XIcon className="h-4 w-4" />
+          </button>
+        )}
+
+        <AddDatabaseDialog>
+          <button
+            type="button"
+            className="float-right border-b px-2 m-0 h-full hover:bg-accent hover:text-accent-foreground"
+          >
+            <PlusIcon className="h-4 w-4" />
+          </button>
+        </AddDatabaseDialog>
+      </div>
+
+      <CommandList className="flex flex-col">
+        {dataConnections.map((connection) => (
+          <Engine
+            key={connection.name}
+            connection={connection}
+            hasChildren={connection.databases.length > 0}
+          >
+            {connection.databases.map((database) => (
+              <DatabaseItem
+                key={database.name}
+                database={database}
+                hasSearch={hasSearch}
+              >
+                <SchemaList
+                  schemas={database.schemas}
+                  engineName={connection.name}
+                  databaseName={database.name}
+                  hasSearch={hasSearch}
+                  searchValue={searchValue}
+                />
+              </DatabaseItem>
+            ))}
+          </Engine>
+        ))}
+
+        {dataConnections.length > 0 && tables.length > 0 && (
+          <DatasourceLabel>
+            <PythonIcon className="h-4 w-4 text-muted-foreground" />
+            <span>Python</span>
+          </DatasourceLabel>
+        )}
+        {tables.length > 0 && (
+          <TableList tables={tables} searchValue={searchValue} />
+        )}
+      </CommandList>
+    </Command>
+  );
+};
+
+const Engine: React.FC<{
+  connection: DataSourceConnection;
+  children: React.ReactNode;
+  hasChildren?: boolean;
+}> = ({ connection, children, hasChildren }) => {
+  const hasEngine = connection.databases.length > 0;
+  // If the connection has no engine, it's the internal duckdb engine
+  const engineName = hasEngine
+    ? connection.databases[0]?.engine || "In-Memory"
+    : connection.name;
+
+  return (
+    <>
+      <DatasourceLabel>
```

Comment:
maybe in future, we could add a super small refresh icon that only appears on hover (using `hover-action` and `hover-action-parent` class names), that would refresh the tables